File: app/src/main/java/com/norvera/confession/utils/DividerDecoration.java
package com.norvera.confession.utils;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Rect;
import android.util.TypedValue;
import android.view.View;

import com.norvera.confession.ui.prayers.SectionedAdapter;

import androidx.recyclerview.widget.RecyclerView;

public class DividerDecoration extends RecyclerView.ItemDecoration {

    private final Paint mPaint;
    private final int mHeightDp;

    public DividerDecoration(Context context) {
        this(context, Color.argb((int) (255 * 0.2), 255, 255, 255), 1f);
    }

    private DividerDecoration(Context context, int color, float heightDp) {
        mPaint = new Paint();
        mPaint.setStyle(Paint.Style.FILL);
        mPaint.setColor(color);
        mHeightDp = (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, heightDp, context.getResources().getDisplayMetrics());
    }

    @Override
    public void getItemOffsets(Rect outRect, View view, RecyclerView parent, RecyclerView.State state) {
        if (hasDividerOnBottom(view, parent, state)) {
            outRect.set(0, 0, 0, mHeightDp);
        } else {
            outRect.setEmpty();
        }
    }

    private boolean hasDividerOnBottom(View view, RecyclerView parent, RecyclerView.State state) {
        int position = ((RecyclerView.LayoutParams) view.getLayoutParams()).getViewAdapterPosition();
        return position < (state.getItemCount() -1) && parent.getAdapter().getItemViewType(position) != SectionedAdapter.HEADER_VIEW_TYPE
                && parent.getAdapter().getItemViewType(position + 1) != SectionedAdapter.HEADER_VIEW_TYPE;
    }

    @Override
    public void onDraw(Canvas c, RecyclerView parent, RecyclerView.State state) {
        for (int i = 0; i < parent.getChildCount(); i++) {
            View view = parent.getChildAt(i);
            if (hasDividerOnBottom(view, parent, state)) {
                c.drawRect(view.getLeft(), view.getBottom(), view.getRight(), view.getBottom() + mHeightDp, mPaint);
            }
        }
    }
}

File: app/src/main/java/com/norvera/confession/utils/DividerDecoration.kt
package com.norvera.confession.utils

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Rect
import android.util.TypedValue
import android.view.View

import com.norvera.confession.ui.prayers.SectionedAdapter

import androidx.recyclerview.widget.RecyclerView

class DividerDecoration @JvmOverloads constructor(
    context: Context,
    color: Int = Color.argb(
        (255 * 0.2).toInt(),
        255,
        255,
        255
    ),
    heightDp: Float = 1f
) : RecyclerView.ItemDecoration() {

    private val mPaint: Paint = Paint()
    private val mHeightDp: Int

    init {
        mPaint.style = Paint.Style.FILL
        mPaint.color = color
        mHeightDp = TypedValue.applyDimension(
            TypedValue.COMPLEX_UNIT_DIP,
            heightDp,
            context.resources.displayMetrics
        ).toInt()
    }

    override fun getItemOffsets(
        outRect: Rect,
        view: View,
        parent: RecyclerView,
        state: RecyclerView.State
    ) {
        if (hasDividerOnBottom(view, parent, state)) {
            outRect.set(0, 0, 0, mHeightDp)
        } else {
            outRect.setEmpty()
        }
    }

    private fun hasDividerOnBottom(
        view: View,
        parent: RecyclerView,
        state: RecyclerView.State
    ): Boolean {
        val position = (view.layoutParams as RecyclerView.LayoutParams).viewAdapterPosition
        return (position < state.itemCount - 1 && parent.adapter!!.getItemViewType(position) != SectionedAdapter.HEADER_VIEW_TYPE
                && parent.adapter!!.getItemViewType(position + 1) != SectionedAdapter.HEADER_VIEW_TYPE)
    }

    override fun onDraw(c: Canvas, parent: RecyclerView, state: RecyclerView.State) {
        for (i in 0 until parent.childCount) {
            val view = parent.getChildAt(i)
            if (hasDividerOnBottom(view, parent, state)) {
                c.drawRect(
                    view.left.toFloat(),
                    view.bottom.toFloat(),
                    view.right.toFloat(),
                    (view.bottom + mHeightDp).toFloat(),
                    mPaint
                )
            }
        }
    }
}
